### PR TITLE
refactor: inject engine clock

### PR DESCRIPTION
**Pull request by lenaschoenburg**
Before this PR, the engine was directly using `ActorClock` in various places to get the current time. The implicit assumption was that the engine would always run in an actor thread where such an `ActorClock` is available.

This PR refactors (mostly the engine) such that a `ControllableStreamClock` is injected as clock. It is injected by the `StreamProcessor` which takes the current `ActorClock` as source. Because we still use the same clock, no behavior changes are expected.

This is useful for two reasons:
1. Being able to run the engine outside of an actor thread.
2.  Injecting the clock gives us more control over the clock (e.g. for testing)

Having the engine not rely on the actor scheduler was a long-standing goal. An architecture test that was supposed to verify this goal is fixed now, proving that engine production code no longer depends on the actor scheduler.

relates to #21051

### CODE REVIEW

**Comment by lenaschoenburg** [comment]
For https://github.com/camunda/product-hub/issues/2379, we'd want to inject our own interface instead of `InstantSource` such that the engine can also _modify_ the current time.

File: zeebe/engine/pom.xml
Comment (npepinpe) [inline zeebe/engine/pom.xml:120]: :question: I thought we had removed all dependencies to the scheduler for the engine? :thinking:

File: zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/api/StreamClock.java
Comment (npepinpe) [inline zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/api/StreamClock.java:47]: :question: This implies that if I say, `offset(5)`, then `offset(5)`, nothing will change? In the previous controlled clock this would actually keep incrementing by 5, as it was always a relative offset. Is this the behavior we want?

File: zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/api/StreamClock.java
Comment (npepinpe) [inline zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/api/StreamClock.java:71]: :wrench: Cool use of sealed types, but I'm wondering if it's required in this instance. It's unlikely to be extended, so right now the UX is:

```java
clock.applyModification(Modification.pinAt(Instant.now()));
```

versus

```java
clock.pinAt(Instant.now());
```

While it makes it easier to extend, and easier if we keep the behavior of overriding previous modifications, the actual API is arguably worse here then just having explicit methods to pin/reset/add offset. I see you already added `reset`, which implies to me that you might have realized (i.e. `applyModification(Modification.none())` is not intuitive to reset).

File: zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/ControllableStreamClockImpl.java
Comment (npepinpe) [inline zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/ControllableStreamClockImpl.java:18]: :+1: Definitely easier to read in general than the current `ControlledActorClock`

File: zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/UncontrolledStreamClock.java
Comment (npepinpe) [inline zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/UncontrolledStreamClock.java:17]: :thought_balloon: Likely if we did not implement `InstantSource` but our own type like `EpochClock`, then we wouldn't need an implementation here but could return a constant lambda, e.g. `() -> InstantSource.system().millis()`.

**Review by npepinpe — APPROVED** [review]
:+1: Some suggestions, no blockers, no need for another review (but of course if you'd like). Please address comments before merging :)

:wrench: I would propose also not using `InstantSource`, but something like Agrona's `EpochClock` - we can also just write our own in `util` if we don't want to couple ourselves (though we already are :smile:) to them. My reasoning here is that we typically only work with epoch timestamps, and `InstantSource` also lets you get things like an `Instant`. Typically for the high performance stuff we'd want to avoid that, so it might just be simpler to not have it at all. It's a very minor future optimization, so I'm also fine without it.

**Comment by npepinpe** [comment]
One last thing, can we verify if we need the offset thing? If it's really just pin, let's start with that.

File: zeebe/engine/pom.xml
Comment (lenaschoenburg) [reply zeebe/engine/pom.xml:120]: This is a test scoped dependency because various tests make use of the actor scheduler to run the engine, see for example `StreamProcessorRule`.

File: zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/api/StreamClock.java
Comment (lenaschoenburg) [reply zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/api/StreamClock.java:47]: Yes, that's also part of the docstring:
```
/**
 * Modifies the clock by applying the given modification. Previous modifications are overridden
 * and do not stack up.
 *
```

> Is this the behavior we want

I don't really know yet. If not, the desired behavior can be emulated easily with first querying current offset, adding to it and then applying a new one.

File: zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/api/StreamClock.java
Comment (lenaschoenburg) [reply zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/api/StreamClock.java:71]: One benefit from it is that it's a symmetric API with `currentModification` and `applyModification`. Doesn't stop us from adding convenience methods such as `pinAt` that work the same way as `reset`.

Comment (lenaschoenburg) [reply zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/api/StreamClock.java:71]: I've added two new helpers, `pinAt` and `offsetBy`.

File: zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/api/StreamClock.java
Comment (lenaschoenburg) [reply zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/api/StreamClock.java:47]: I've added a convenience method to do that, `stackOffset`.

**Comment by lenaschoenburg** [comment]
> One last thing, can we verify if we need the offset thing? If it's really just pin, let's start with that.

I don't think we can get a better answer than the one we already have which is basically "probably pin is enough". I don't think offset is adding much complexity here so I'd just keep it for now.

File: zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/UncontrolledStreamClock.java
Comment (lenaschoenburg) [reply zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/UncontrolledStreamClock.java:17]: Actually this implementation is already not used, maybe we can just remove it along with the interface `StreamClock`? I don't think we have a usecase where the clock should _not_ be controllable.

File: zeebe/engine/pom.xml
Comment (npepinpe) [reply zeebe/engine/pom.xml:120]: Oh :( I would have hoped even in the tests we didn't :sweat_smile:

File: zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/UncontrolledStreamClock.java
Comment (npepinpe) [reply zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/UncontrolledStreamClock.java:17]: Th only risk is in the future if we want an immutable one :shrug: I could imagine that we will have a feature flag for this, so the question is do we want to make it completely impossible, or just prevent commands from modifying the clock? I would lean towards the safer/cautious approach, but in the end with good test coverage it's likely fine either way.

Comment (lenaschoenburg) [reply zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/UncontrolledStreamClock.java:17]: Okay then let's keep it for now, maybe we can use it to disable the entire feature :+1:

**Comment by npepinpe** [comment]
Maybe I missed it, but I didn't see you address the review comment:

> 🔧 I would propose also not using InstantSource, but something like Agrona's EpochClock - we can also just write our own in util if we don't want to couple ourselves (though we already are 😄) to them. My reasoning here is that we typically only work with epoch timestamps, and InstantSource also lets you get things like an Instant. Typically for the high performance stuff we'd want to avoid that, so it might just be simpler to not have it at all. It's a very minor future optimization, so I'm also fine without it.

**Comment by lenaschoenburg** [comment]
Ah yes, I didn't address it :+1: 
I'd rather continue using `InstantSource` because it's a standardized, easy to use interface that offers more functionality than something like agrona's `EpochClock`.
While it's true that we sometimes use millis since epoch instead of `Instant` or `Duration`, having the ability to get an `Instant` from the clock is still nice.

I also feel like using `InstantSource` does not add any complexity apart from having to implement two methods _if_ the `millis` code path needs to be optimized.

**Pull request merged**
